homeassistant/components/local_todo/todo.py

```diff
@@ -0,0 +1,162 @@
+"""A Local To-do todo platform."""
+
+from collections.abc import Iterable
+import dataclasses
+import logging
+from typing import Any
+
+from ical.calendar import Calendar
+from ical.calendar_stream import IcsCalendarStream
+from ical.store import TodoStore
+from ical.todo import Todo, TodoStatus
+from pydantic import ValidationError
+
+from homeassistant.components.todo import (
+    TodoItem,
+    TodoItemStatus,
+    TodoListEntity,
+    TodoListEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import CONF_TODO_LIST_NAME, DOMAIN
+from .store import LocalTodoListStore
+
+_LOGGER = logging.getLogger(__name__)
+
+
+PRODID = "-//homeassistant.io//local_todo 1.0//EN"
+
+ICS_TODO_STATUS_MAP = {
+    TodoStatus.IN_PROCESS: TodoItemStatus.NEEDS_ACTION,
+    TodoStatus.NEEDS_ACTION: TodoItemStatus.NEEDS_ACTION,
+    TodoStatus.COMPLETED: TodoItemStatus.COMPLETED,
+    TodoStatus.CANCELLED: TodoItemStatus.COMPLETED,
+}
+ICS_TODO_STATUS_MAP_INV = {
+    TodoItemStatus.COMPLETED: TodoStatus.COMPLETED,
+    TodoItemStatus.NEEDS_ACTION: TodoStatus.NEEDS_ACTION,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the local_todo todo platform."""
+
+    store = hass.data[DOMAIN][config_entry.entry_id]
+    ics = await store.async_load()
+    calendar = IcsCalendarStream.calendar_from_ics(ics)
+    calendar.prodid = PRODID
+
+    name = config_entry.data[CONF_TODO_LIST_NAME]
+    entity = LocalTodoListEntity(store, calendar, name, unique_id=config_entry.entry_id)
+    async_add_entities([entity], True)
+
+
+def _todo_dict_factory(obj: Iterable[tuple[str, Any]]) -> dict[str, str]:
+    """Convert TodoItem dataclass items to dictionary of attributes for ical consumption."""
+    result: dict[str, str] = {}
+    for name, value in obj:
+        if name == "status":
+            result[name] = ICS_TODO_STATUS_MAP_INV[value]
+        elif value is not None:
+            result[name] = value
+    return result
+
+
+def _convert_item(item: TodoItem) -> Todo:
+    """Convert a HomeAssistant TodoItem to an ical Todo."""
+    try:
+        return Todo.parse_obj(dataclasses.asdict(item, dict_factory=_todo_dict_factory))
+    except ValidationError as err:
```

Comment:
This looks like the library isn't abstracted properly in this interface. I don't think the library user should know that the library uses pydantic.

Reply:
OK -- we'll want to update [local calendar](https://github.com/home-assistant/core/blob/80b3fec6753c275f056a47187ebad92ebd5cd8fe/homeassistant/components/local_calendar/calendar.py#L182) as well then. *Aside: i chose a different error to raise there given it was about input validation based on different review standards. But here, i'm not sure that the user can actually make this happen given format validation happening upstream on the input so its not expected*

Just to be clear though: The way `ical` uses pydantic is not just an accidental leakage of an internal underlying detail -- the entirely library is built to intentionally make you use data objects that are pydantic data classes (e.g. [Event](https://github.com/allenporter/ical/blob/main/ical/event.py#L61) and the examples https://allenporter.github.io/ical/ical.html#quickstart) are directly using the pydantic constructors and helpers as convenience methods rather than having multiple layers of wrapping. (The objects being used here are a `BaseModel`)

When designing the `ical` library pydantic didn't seem like it was going to randomly introduce an upgrade nightmare of changes given it was recommended by core team members as a defacto data serialization library, but lesson learned.

The smallest possible change is to add method with the same signature as `parse_obj` throws an ical specific exception.  Larger changes could be changing the base objects to no longer by pydantic objects.

Happy to have any library design advice here given that context.

Reply:
I think it's enough to replicate the method and throw an ical specific exception instead of a pydantic exception.

It's not a problem to return BaseModel objects to the user as the user doesn't need to care about the base class. It's just a typed object with attributes as far as the user needs to know.

Reply:
I don't see any pydantic specific interfaces or mentions in the linked library docs. That looks good.

Reply:
Good!